metadata/src/utils/validation.rs

```diff
@@ -195,31 +190,53 @@ fn get_type_def_hash(
 pub fn get_type_hash(
     registry: &PortableRegistry,
     id: u32,
-    visited_ids: &mut HashSet<u32>,
+    cache: &mut HashMap<u32, u32>,
 ) -> [u8; HASH_LEN] {
-    // Guard against recursive types and return a fixed arbitrary hash
-    if !visited_ids.insert(id) {
-        return hash(MAGIC_RECURSIVE_TYPE_VALUE);
+    // Guard against recursive types, with a 2 step caching approach:
+    //    if the cache has an entry for the id, just return a hash derived from it.
+    //    if the type has not been seen yet, mark it with a 0 in the cache and proceed to `get_type_def_hash()`.
+    //        -> During the execution of get_type_def_hash() we might get into get_type_hash(id) again for the original id
+    //            -> in this case the 0 provokes an early return.
+    //        -> Once we return from `get_type_def_hash()` we need to update the entry in the cache.
+    //            -> We just derive some u32 from the hash that was returned and save it.
+    //            -> This makes sure, that different types end up with different cache values.
+    //
+    // Values in the cache can be thought of as a mapping like this:
+    // type id -> not contained           if not seen yet
+    //         -> 0                       if within get_type_hash() calculation for the type
+    //         -> (pseudo-)random u32     if get_type_hash() calculation has completed for the type
+
+    if let Some(short_hash) = cache.get(&id) {
+        return hash(&short_hash.to_be_bytes());
     }
-
+    cache.insert(id, 0);
```

Comment:
I like the idea of caching the hashes like this in general!

There's maybe a small chance that the first 4 bytes of `type_hash` are actually `[0,0,0,0]`, which would make that type look like the special recursive type that leads to an early return.

I also wonder; is it better to store the full `[u8; HASH_LEN]` type hash in the cache? We use more memory with this, but then don't need to `hash` it, and it means that the cache is truly caching the computed hashes for various types rather than caching something that's not actually the original hash we computed for a type.

Combining the two thoughts above, I wonder whether the cache should look like:

```rust
type Cache = HashMap<u32, CachedValue>;

enum CachedValue {
    Recursive,
    Hash([u8; HASH_LEN])
}
```

Where we insert `CachedValue::Recursive` initially and update that to a final hash when done. (if `get_type_hash` looks in the cache and sees a `CachedValue::Recursive` it can return with eg `hash(MAGIC_RECURSIVE_HASH_VALUE)`, which is functionally the same as right now where it would return with `hash(0)`).

I'm suggesting a new enum above over using `Option` just because it would be a little more self-documenting, but either works really!

Reply:
We could definitely do that, thought about something like that as well, it is more secure at the cost of a little bit more memory. I will change it! :)

Reply:
That'd be great, and then perhaps also we could add a test which uses the cache across a couple of invocations of hashing the same type to comfirm that indeed you get the same hash back both times (first time caches the hash, second time just returns it straight off). This should even work if the type has a recursive definition in which is cool :)